Userland/Applications/PixelPaint/Tools/EraseTool.cpp

```diff
@@ -146,16 +146,52 @@ NonnullRefPtr<Gfx::Bitmap> EraseTool::build_cursor()
     m_scale_last_created_cursor = m_editor ? m_editor->scale() : 1;
     int scaled_size = size() * m_scale_last_created_cursor;
 
+    // If we have an ImageEditor the scaled_size should not excede the longest size of the ImageEditor (width or height)
+    if (m_editor) {
+        int max_dimension = max(m_editor->width(), m_editor->height());
+        if (scaled_size > max_dimension * 2) {
+            // scaled_size needs to change in order to draw the cursor in the correct location.
+            scaled_size = max_dimension * 2;
+        }
+    }
+    scaled_size = max(scaled_size, 1);
     NonnullRefPtr<Gfx::Bitmap> new_cursor = Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, Gfx::IntSize(scaled_size, scaled_size)).release_value_but_fixme_should_propagate_errors();
 
     Gfx::IntRect rect { 0, 0, scaled_size, scaled_size };
     Gfx::Painter painter { new_cursor };
 
-    painter.draw_rect(rect, Color::LightGray);
-    painter.draw_line({ scaled_size / 2 - 5, scaled_size / 2 }, { scaled_size / 2 + 5, scaled_size / 2 }, Color::LightGray, 3);
-    painter.draw_line({ scaled_size / 2, scaled_size / 2 - 5 }, { scaled_size / 2, scaled_size / 2 + 5 }, Color::LightGray, 3);
-    painter.draw_line({ scaled_size / 2 - 5, scaled_size / 2 }, { scaled_size / 2 + 5, scaled_size / 2 }, Color::MidGray, 1);
-    painter.draw_line({ scaled_size / 2, scaled_size / 2 - 5 }, { scaled_size / 2, scaled_size / 2 + 5 }, Color::MidGray, 1);
+    // If no ImageEditor is present, we cannot bind the rect within the editor. Then we will just draw the rect as normal.
+
+    auto half_scaled_size = scaled_size / 2;
+    if (!m_editor) {
+        painter.draw_rect(rect, Color::LightGray);
+    } else {
+        if (m_current_tool_position.x() >= (half_scaled_size) && (m_current_tool_position.x() + half_scaled_size) <= m_editor->width() && m_current_tool_position.y() >= (half_scaled_size) && (m_current_tool_position.y() + half_scaled_size) <= m_editor->height()) {
+            painter.draw_rect(rect, Color::LightGray);
+        } else {
+            int left_overlap = (m_current_tool_position.x() < (half_scaled_size)) ? ((half_scaled_size)-m_current_tool_position.x()) : 0;
```

Comment:
Is that intersection check something that could be done with the Rect.intersect function? I thing that would be more readable.

Overall i think that generating something like a "dynamic" cursor is not as it's intended to be used as you currently always generate and draw a new cursor bitmap - not only when the dimensions change.

I belive the better way would be to utilize the "on_second_paint" function to draw the cursor and maybe to get rid of the overdrawing outside the editor window. (That was my faulft - i kind of like it to see the overlaping cursor and implented it that way... bit it's fine to change that.)

Reply:
Yes I agree that a dynamic cursor doesn't make much sense because `m_cursor` itself  is constant and doesn't allow us to change how it is painted without rebuilding the cursor.

Do you mind elaborating on how `on_second_paint` could be used to get rid of the overdrawing. I am having a bit of trouble understanding when/how that function is called, and what is it's intended use

Reply:
The on_second_paint function is a virtual fuction that the Tool or Widget class provides and is called after the normal paint has occured to provide the ability to draw a "overlay" in a second paint pass.
You can have a look at the GradientTool that does make use of the on_second_paint function and does clip the painter to the size of the editor widget.

Reply:
Are we able to paint the cursor outside of build_cursor() when m_cursor is expected to be constant?

Reply:
Yes that is possible, if you are changing the logic to draw the cursor with the "on_second_paint" function you would not need to create a cursor bitmap anymore, instead the logic that now draws the cursor onto the bitmap  is used in the new function to directly draw the cursor into the editor window.
The cursor function could then return something like StandardCursor::None or Hidden, not sure what the best would be.